plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java

```diff
@@ -89,7 +120,20 @@ public void backupWorkspace(final String workspaceId,
     public void backupWorkspaceAndCleanup(final String workspaceId,
                                           final String srcPath,
                                           final String srcAddress) throws ServerException {
-        backupWorkspace(workspaceId, srcPath, srcAddress, true);
+        ReentrantLock lock = workspacesBackupLocks.get(workspaceId);
+        if (lock != null) {
+            lock.lock();
+            if (workspacesBackupLocks.get(workspaceId) == null) {
+                // should never happen
+                LOG.error("Backup with cleanup of the workspace {} was invoked several times simultaneously", workspaceId);
+            }
+            try {
```

Comment:
I think It is convenient to put try block right after locking. Please do that to not confuse possible code readers.